File: src/main/java/at/fhhagenberg/sqe/api/FloorButtonServiceImpl.java
package at.fhhagenberg.sqe.api;

import org.jetbrains.annotations.NotNull;
import sqelevator.IElevator;
import at.fhhagenberg.sqe.entity.FloorButton;
import com.google.inject.Inject;
import java.rmi.RemoteException;
import java.util.ArrayList;
import java.util.List;

public class FloorButtonServiceImpl implements FloorButtonService {

    private final IElevator elevatorControl;

    @Inject
    public FloorButtonServiceImpl(
            IElevator elevatorControl
    ) {
        this.elevatorControl = elevatorControl;
    }

    @Override
    @NotNull
    public List<FloorButton> getAll(int elevatorNumber) throws RemoteException {
        int totalNumberOfElevators = elevatorControl.getElevatorNum();
        int totalNumberOfFloors = elevatorControl.getFloorNum();
        if (totalNumberOfFloors < 0) {
            totalNumberOfFloors = 0;
        }
        List<FloorButton> buttons = new ArrayList<>(totalNumberOfFloors);

        if (elevatorNumber >= 0 && elevatorNumber < totalNumberOfElevators) {
            for (int floorNumber = 0; floorNumber < totalNumberOfFloors; floorNumber++) {
                FloorButton button = get(elevatorNumber, floorNumber);
                if (button != null) {
                    buttons.add(button);
                }
            }
        }
        return buttons;
    }

    @Override
    public FloorButton get(int elevatorNumber, int floorNumber) throws RemoteException {
        int totalNumberOfElevators = elevatorControl.getElevatorNum();
        int totalNumberOfFloors = elevatorControl.getFloorNum();

        if (elevatorNumber >= 0 && elevatorNumber < totalNumberOfElevators && floorNumber >= 0 && floorNumber < totalNumberOfFloors) {
            boolean isActive = elevatorControl.getElevatorButton(elevatorNumber, floorNumber);
            return new FloorButton(elevatorNumber, floorNumber, isActive);
        }
        return null;
    }
}

File: src/main/kotlin/at/fhhagenberg/sqe/api/FloorButtonServiceImpl.kt
package at.fhhagenberg.sqe.api

import at.fhhagenberg.sqe.entity.FloorButton
import com.google.inject.Inject
import sqelevator.ConnectableIElevator
import java.rmi.RemoteException
import java.util.*

class FloorButtonServiceImpl @Inject constructor(
        private val elevatorControl: ConnectableIElevator
) : FloorButtonService {
    @Throws(RemoteException::class)
    override fun getAll(elevatorNumber: Int): List<FloorButton> {
        val totalNumberOfElevators = elevatorControl.elevatorNum
        val totalNumberOfFloors = elevatorControl.floorNum
        val floorButtons: MutableList<FloorButton> = ArrayList(totalNumberOfFloors.coerceAtLeast(0))
        if (elevatorNumber in 0 until totalNumberOfElevators) {
            for (floorNumber in 0 until totalNumberOfFloors) {
                floorButtons.add(get(elevatorNumber, floorNumber)!!)
            }
        }
        return floorButtons
    }

    @Throws(RemoteException::class)
    override fun get(elevatorNumber: Int, floorNumber: Int): FloorButton? {
        val totalNumberOfElevators = elevatorControl.elevatorNum
        val totalNumberOfFloors = elevatorControl.floorNum
        return if (elevatorNumber in 0 until totalNumberOfElevators && floorNumber in 0 until totalNumberOfFloors) {
            val isActive = elevatorControl.getElevatorButton(elevatorNumber, floorNumber)
            FloorButton(elevatorNumber, floorNumber, isActive)
        } else null
    }
}